From 9080bb10e37dd2311da227650f33d7fa831b2de6 Mon Sep 17 00:00:00 2001
From: Allen Ye <allen.ye@mediatek.com>
Date: Mon, 24 Feb 2025 16:25:48 +0800
Subject: [PATCH 67/95] mtk: mt76: mt7996: Force to update power after csa
 count down

Force to update txpower sku table after csa count down.

mac80211 will not update power in __ieee80211_recalc_txpower() if the
maximum txpower of the both channels are the same.

Signed-off-by: Allen Ye <allen.ye@mediatek.com>
---
 mt7996/main.c | 15 +++++++++++----
 1 file changed, 11 insertions(+), 4 deletions(-)

diff --git a/mt7996/main.c b/mt7996/main.c
index 1cfc75ef..cdc56061 100644
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -645,6 +645,9 @@ int mt7996_set_channel(struct mt76_phy *mphy)
 		if (ret)
 			goto out;
 	} else if (mphy->chanctx && mphy->chanctx->state == MT76_CHANCTX_STATE_SWITCH) {
+		u8 delta;
+		int current_txpower;
+
 		if (mphy->chanctx->has_ap && phy->pp_mode == PP_USR_MODE) {
 			ret = mt7996_mcu_set_pp_en(phy, PP_USR_MODE,
 						   mphy->main_chandef.punctured);
@@ -654,6 +657,14 @@ int mt7996_set_channel(struct mt76_phy *mphy)
 			ret = mt7996_mcu_set_pp_sta_dscb(phy, &mphy->main_chandef,
 							 omac_idx);
 		}
+		if (ret)
+			goto out;
+
+		delta = mt76_tx_power_path_delta(hweight16(mphy->chainmask));
+		current_txpower = DIV_ROUND_UP(mphy->txpower_cur + delta, 2);
+		ret = mt7996_mcu_set_txpower_sku(phy, current_txpower);
+		if (ret)
+			goto out;
 	}
 
 	if (phy->dev->cal) {
@@ -675,10 +686,6 @@ int mt7996_set_channel(struct mt76_phy *mphy)
 	if (ret)
 		goto out;
 
-	ret = mt7996_mcu_set_txpower_sku(phy);
-	if (ret)
-		goto out;
-
 	ret = mt7996_dfs_init_radar_detector(phy);
 	mt7996_mac_cca_stats_reset(phy);
 
-- 
2.45.2

